esp-hal/src/dma/buffers.rs

```diff
@@ -1562,3 +1562,391 @@ impl DerefMut for DmaLoopBuf {
+            unsafe { aligned_data.cast::<u8>().byte_add(chunk) },
+            aligned_data.len() - chunk,
+        );
+    }
+
+    // Align end
+    if tail_to_copy > 0 {
+        let copy_buffer = unwrap!(copy_buffer_iter.next());
+        let buffer = copy_buffer.insert(ManualWritebackBuffer::new(
+            unsafe { data.cast().byte_add(data.len() - tail_to_copy) },
+            tail_to_copy,
+        ));
+
+        let Some(descriptor) = desciptor_iter.next() else {
+            return consumed;
+        };
+        descriptor.set_size(tail_to_copy);
+        descriptor.buffer = buffer.buffer_ptr();
+        consumed += tail_to_copy;
+    }
+
+    consumed
+}
+
+#[cfg(psram_dma)]
+struct DescriptorChainingIter<'a> {
+    /// index of the next element to emit
+    index: usize,
+    descriptors: &'a mut [DmaDescriptor],
+}
+#[cfg(psram_dma)]
+impl<'a> DescriptorChainingIter<'a> {
+    fn new(descriptors: &'a mut [DmaDescriptor]) -> Self {
+        Self {
+            descriptors,
+            index: 0,
+        }
+    }
+
+    fn next(&mut self) -> Option<&'_ mut DmaDescriptor> {
+        if self.index == 0 {
+            self.index += 1;
+            self.descriptors.get_mut(0)
+        } else if self.index < self.descriptors.len() {
+            let index = self.index;
+            self.index += 1;
+
+            // Grab a pointer to the current descriptor.
+            let ptr = &raw mut self.descriptors[index];
+
+            // Link the descriptor to the previous one.
+            self.descriptors[index - 1].next = ptr;
+
+            // Reborrow the pointer so that it doesn't get invalidated by our continued use of the
+            // descriptor reference.
+            Some(unsafe { &mut *ptr })
+        } else {
+            None
+        }
+    }
+}
+
+unsafe impl DmaRxBuffer for UnsafeRxBuffer<'_> {
+    type View = Self;
+
+    fn prepare(&mut self) -> Preparation {
+        Preparation {
+            start: self.descriptors.head(),
+            direction: TransferDirection::In,
+            burst_transfer: BurstConfig::DEFAULT,
+            check_owner: None,
+            auto_write_back: true,
+            #[cfg(psram_dma)]
+            accesses_psram: self.in_psram,
+        }
+    }
+
+    fn into_view(self) -> Self::View {
+        self
+    }
+
+    fn from_view(view: Self::View) -> Self {
+        view
+    }
+}
+
+#[cfg(psram_dma)]
+const MIN_LAST_DMA_LEN: usize = if cfg!(esp32s2) { 5 } else { 1 };
+#[cfg(psram_dma)]
+const BUF_LEN: usize = 16 + 2 * (MIN_LAST_DMA_LEN - 1); // 2x makes aligning short buffers simpler
+
+/// PSRAM helper. DMA can write data of any alignment into this buffer, and it can be written by
+/// the CPU back to PSRAM.
+#[cfg(psram_dma)]
+pub(crate) struct ManualWritebackBuffer {
+    dst_address: NonNull<u8>,
+    buffer: [u8; BUF_LEN],
+    n_bytes: u8,
+}
+
+#[cfg(psram_dma)]
+impl ManualWritebackBuffer {
+    pub fn new(ptr: NonNull<u8>, len: usize) -> Self {
```

Comment:
Could go with fat pointers for these
```suggestion
    pub fn new(ptr: NonNull<[u8]>) -> Self {
```